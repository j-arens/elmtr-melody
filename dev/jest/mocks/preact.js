module.exports = {
    unmountComponent: jest.fn(),
};
