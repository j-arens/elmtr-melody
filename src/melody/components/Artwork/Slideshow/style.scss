.Slideshow {
    display: flex;
    width: 100%;
    height: 100%;
    padding: 0;
    margin: 0;
    list-style: none !important;

    &__slide {
        flex: 1 0 100%;
        height: 100%;
        padding: 0;
        margin: 0;
    }

    &__image {
        width: 100%;
        height: 100%;
    }
}
