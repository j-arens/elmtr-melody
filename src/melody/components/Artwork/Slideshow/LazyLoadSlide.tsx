import { WithOptionalClassName } from '@components/type';
import { h } from 'preact';

interface Props extends WithOptionalClassName {
    currentTrack: number;
    totalTracks: number;
    index: number;
    className?: string;
    children?: JSX.Element[];
}

const EmptySlide = (className: string): JSX.Element =>
    <li class={className} />;

const LoadedSlide = (
    className: string,
    children: JSX.Element[],
): JSX.Element => (
    <li class={className}>
        {children}
    </li>
);

export default ({
    currentTrack,
    totalTracks,
    index,
    className = '',
    children,
}: Props) => {
    const loadLast = currentTrack === 0 && index === (totalTracks - 1);
    const loadFirst = currentTrack === (totalTracks - 1) && index === 0;
    const aboveRange = index > (currentTrack + 1);
    const belowRange = index < (currentTrack - 1);

    if (loadLast || loadFirst) {
        return LoadedSlide(className, children);
    }

    if (aboveRange || belowRange) {
        return EmptySlide(className);
    }

    return LoadedSlide(className, children);
};
