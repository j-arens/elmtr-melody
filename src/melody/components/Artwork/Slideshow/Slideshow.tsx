import ResponsiveBgImage from '@components/Artwork/ResponsiveBgImage';
import { WithOptionalClassName } from '@components/type';
import { Track } from '@redux/type';
import { h } from 'preact';
import LazyLoadSlide from './LazyLoadSlide';
const s = require('./style.scss');

export interface StateProps {
    tracks: Track[];
    currentTrack: number;
}

interface OwnProps extends WithOptionalClassName {
    artworkClassName?: string;
}

type Props = StateProps & OwnProps;

export default ({
    tracks,
    currentTrack,
    className = '',
    artworkClassName = '',
}: Props) => (
    <ul
        class={`${s.Slideshow} ${className}`}
        style={{
            transform: `translate3d(${currentTrack * -100}%, 0px, 0px)`,
        }}
    >
        {tracks.map((track, i) => (
            <LazyLoadSlide
                index={i}
                totalTracks={tracks.length}
                className={s.Slideshow__slide}
                currentTrack={currentTrack}
            >
                <ResponsiveBgImage
                    className={`${s.Slideshow__image} ${artworkClassName}`}
                    artwork={track.artwork}
                />
            </LazyLoadSlide>
        ))}
    </ul>
);
