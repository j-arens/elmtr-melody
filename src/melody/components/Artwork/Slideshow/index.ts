import { WithOptionalClassName } from '@components/type';
import { State } from '@redux/type';
import { connect } from 'preact-redux';
import Slideshow, { StateProps } from './Slideshow';

const mapStateToProps = (state: State) => ({
    tracks: state.audio.tracks,
    currentTrack: state.audio.currentTrack,
});

export default connect<StateProps, {}, WithOptionalClassName>(mapStateToProps)(Slideshow);
