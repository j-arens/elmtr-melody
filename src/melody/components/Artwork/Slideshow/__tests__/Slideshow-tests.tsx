import { h } from 'preact';
import { render } from 'preact-render-to-string';
import Slideshow from '../Slideshow';

describe('Slideshow', () => {
    let tracks;

    beforeEach(() => {
        tracks = [{
            source_url: '',
            download_url: '',
            artwork: {},
            media_details: {},
            attributes: {},
        }];
    });

    it('renders', () => {
        const component = render(
            <Slideshow
                className="yas-marina-blue"
                currentTrack={0}
                tracks={tracks}
            />,
        );
        expect(component).toMatchSnapshot();
    });

    it('translates based on the current track', () => {
        const component = render(
            <Slideshow
                className="yas-marina-blue"
                currentTrack={5}
                tracks={tracks}
            />,
        );
        expect(component).toMatchSnapshot();
    });
});
