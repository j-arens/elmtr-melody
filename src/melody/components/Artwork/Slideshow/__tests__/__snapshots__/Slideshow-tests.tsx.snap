// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slideshow renders 1`] = `
<ul class="Slideshow yas-marina-blue" data-melody-slideshow style="transform: translate3d(0%, 0px, 0px);">
  <li class="Slideshow__slide">
    <figure class="Slideshow__image" data-melody-artwork style="background-image: url();"></figure>
  </li>
</ul>
`;

exports[`Slideshow translates based on the current track 1`] = `
<ul class="Slideshow yas-marina-blue" data-melody-slideshow style="transform: translate3d(-500%, 0px, 0px);">
  <li class="Slideshow__slide"></li>
</ul>
`;
